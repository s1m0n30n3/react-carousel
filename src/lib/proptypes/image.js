import PropTypes from "prop-types";

import { imageBehaviorProps, effectProps } from "./common";

export const imageProptypes = {
  imagesLength: PropTypes.number,
  currentIndex: PropTypes.number,
  index: PropTypes.number,
  imageBehavior: PropTypes.oneOf(imageBehaviorProps),
  effect: PropTypes.oneOf(effectProps),
  effectSpeed: PropTypes.number,
  imageBlur: PropTypes.shape({
    blur: PropTypes.bool,
    blurAmount: PropTypes.number,
  }),
  images: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      src: PropTypes.string,
      alt: PropTypes.string,
    })
  ),
};
