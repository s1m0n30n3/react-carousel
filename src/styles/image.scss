.image--base {
    padding: 0;
    margin: 0;
    display: inline-flex;
    height: inherit;
    justify-content: center;
    align-items: center;
    overflow: hidden;
    transform: translate3d(0,0,0);
    img {
        user-select: none;
        object-fit: contain;
    }
}

.wrapper--fade-effect {
    width: max-content;
    overflow: hidden;
}
.wrapper--slide-effect {
    width: 100%;
    overflow: hidden;
    position: relative;
}

/* Effects */
.image--fade-effect {
    transition-timing-function: ease-in-out;
    transition-property: opacity;
    position: absolute;
}

.image--inOutSlide-effect {
    transition-timing-function: ease-in-out;
    transition-property: transform;
}

/* Image Formats */
.image--full-format,
.image--contained-format {
    width: 100%;
    height: 100%;
    img {
        width: inherit;
        height: inherit;
    }
}

.image--full-format img {
    object-fit: cover;
}

.image--contained-format img {
    object-fit: contain;
}