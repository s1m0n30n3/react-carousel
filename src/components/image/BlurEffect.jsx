import React from "react";

import { useBlurEffect } from "hooks";
import { blurEffectPropTypes } from "proptypes/index";

import { base } from "styles/blur-effect.module.css";

export const BlurEffect = ({ effects, blur, blurAmount, src, ...props }) => {
  const blurProps = useBlurEffect({ blur, blurAmount, src });

  return (
    <div {...props} className={base} style={{ ...effects, ...blurProps }} />
  );
};

BlurEffect.propTypes = blurEffectPropTypes;

BlurEffect.displayName = "Gallery.BlurEffect";
