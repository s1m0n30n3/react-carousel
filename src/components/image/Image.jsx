import React from "react";

import { classNames } from "helpers";
import { ImageBehavior, TransitionEffect } from "constants";
import { imageProptypes } from "proptypes/index";
import { useImageTransition } from "hooks";

import {
  wrapperFade,
  wrapperSlide,
  base,
  fade,
  full,
  contained,
  inOutSlide,
} from "styles/image.module.css";

import { BlurEffect } from "./BlurEffect";

export const Image = ({
  effect,
  index,
  currentIndex,
  images,
  imageBehavior,
  effectSpeed,
  imageBlur,
  ...props
}) => {
  const isFullWidth = imageBehavior === ImageBehavior.FULL;
  const isContained = imageBehavior === ImageBehavior.CONTAINED;
  const isFadeEffect = effect === TransitionEffect.FADE;
  const isInOutSlideEffect = effect === TransitionEffect.IN_OUT_SLIDE;
  const isCurrentSlide = index === currentIndex;

  const effects = useImageTransition({ isCurrentSlide, effectSpeed });

  return (
    <div role="listitem" className={isFadeEffect ? wrapperFade : wrapperSlide}>
      {!isFullWidth && !!imageBlur.blur && (
        <BlurEffect
          effects={effects[effect]}
          src={images[index]?.src}
          {...imageBlur}
        />
      )}
      <figure
        {...props}
        data-cy={`image-slide-${index}`}
        className={classNames([
          base,
          isFadeEffect && fade,
          isFullWidth && full,
          isContained && contained,
          isInOutSlideEffect && inOutSlide,
          isCurrentSlide && "image-selected",
        ])}
        style={{ ...effects[effect] }}
      />
    </div>
  );
};

Image.propTypes = imageProptypes;

Image.displayName = "Gallery.Image";
