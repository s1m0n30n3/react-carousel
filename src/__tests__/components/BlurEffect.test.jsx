import React from "react";
import renderer from "react-test-renderer";
import { describe, expect, test } from "vitest";

import { BlurEffect } from "../../components/image/BlurEffect";

describe("BlurEffect", () => {
  test("renders correctly without props", () => {
    const component = renderer.create(<BlurEffect />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
