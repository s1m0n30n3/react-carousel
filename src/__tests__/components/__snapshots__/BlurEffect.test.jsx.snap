// Vitest Snapshot v1

exports[`BlurEffect > renders correctly without props 1`] = `
<div
  style={
    {
      "backgroundImage": "initial",
      "filter": "initial",
      "height": "100%",
      "left": 0,
      "top": 0,
      "width": "100%",
      "zIndex": 0,
    }
  }
/>
`;
